README
Slippy now uses the QAAPI (part of /QuickAuctions/modules/scan.lua) to get scan data from Quick Auctions 3.
The version of QA3 that contains this API will not go live until WoW client patch 3.3.3.

In order to allow Slippy to use the QAAPI before it goes live, you must make the following modifications
to your version of Quick Auctions:

1. Navigate to the QuickAuctions directory inside your /World of Warcraft/Interface/Addons folder.
2. Open up the modeules folder.
3. Open scan.lua in an editor (Notepad will work fine).
4. Add the following code to the very end of the file:

-- Simple data API
QAAPI = {}
function QAAPI:GetData(itemLink)
	if( not itemLink ) then return end
	itemLink = QuickAuctions:GetSafeLink(itemLink)
 
	return auctionData[itemLink]
end


5. File->Save to save the changes. You should now see "Loaded Slippy v1.5 and QAAPI successfully!"
and all QAAPI features should be enabled and work properly when you log into wow with Slippy enabled.


If you encounter any errors or you have performed these steps and are still getting a warning message,
please leave a comment detailing your error / problems at:

wow.curse.com/downloads/wow-addons/details/slippy.aspx

ChangeLog.txt
v1.5	-Updated the core.lua file to support the new data tables.
	-Rewrote many functions in core.lua to make it more compact and hopefully easier to read.
	-Made all of the old data tables local inside of data.lua to finalize
	the conversion to the new slippyData table.
	-Incorperated the first version of the new GUI!
v1.4	-Quick Auctions API implemented!
	-Wrote functions to deal with the QAAPI as well as deal with itemLinks.
	-Added a new row to all the data arrays where the itemID of the scroll is stored.
	-Added slippyData which will eventually replace the 9 individual data tables. For now,
	it simply contains the 9 data tables which allows them to be acessed in the same manner
	as before as well as through the slippyData array without conflict.
	-Converted everything inside of the core.lua file to use slippyData instead of the old arrays.
	-Matched the order of the arrays in data.lua to the rest of the program for consistency.
	-Added a warning if QAAPI is not loaded (currently does not do anything besides print the warning).

v1.3	-Multiple adjustments to the core.lua file to improve the code.
	-More variables defined as locals and some variables renamed / moved as a result.

v1.2	-Added new rows for the arrays in data.lua to handle extra materials. As a result,
	everything that used these arrays was updated (mainly core.lua).
	-Stored the totals in the DB. As a result everything that used the totals array was updated (mainly gui.lua).
	-Improved the slash commands greater by making "/sl" print out list of slash commands.
	-Added an option to enable / disabled vellums.
	-Cleaned up the core.lua code by putting the Get and Set function for the options in the options function.
	-Moved the default profile values to the data.lua file.

v1.1 	-Fixed a bug where glove enchants were not being reset when the "Reset Craft Queue" button was pressed.
	-Greatly improved the documentation of the code by adding comments throughout.
	-Cleaned up the code overall by getting rid of unused/unncessary functions/variables.

v1.0 	-First Release Version!

gui.lua
function Slippy:FrameToggle()
	local totalCost = "0"
	
	local function DrawGroup1(container)
		local desc = AceGUI:Create("Label")
		desc:SetText("This is Tab 1")
		desc:SetFullWidth(true)
		container:AddChild(desc)
		 
		local button = AceGUI:Create("Button")
		button:SetText("Tab 1 Button")
		button:SetWidth(200)
		container:AddChild(button)
		button:SetCallback("OnClick", function() self.frame:hide() end)
	end

	-- function that draws the widgets for the second tab
	local function DrawGroup2(container)
		local desc = AceGUI:Create("Label")
		desc:SetText("This is Tab 2")
		desc:SetFullWidth(true)
		container:AddChild(desc)
		
		local button = AceGUI:Create("Button")
		button:SetText("Tab 2 Button")
		button:SetWidth(200)
		container:AddChild(button)
	end

	-- Callback function for OnGroupSelected
	local function SelectGroup(container, event, group)
		container:ReleaseChildren()
		if group == "tab1" then
			DrawGroup1(container)
		elseif group == "tab2" then
			DrawGroup2(container)
		end
	end

	-- Create the frame container
	frame = AceGUI:Create("Frame")
	frame:SetTitle("Example Frame")
	frame:SetStatusText(totalCost)
	frame:SetCallback("OnClose", function(widget) AceGUI:Release(widget) end)
	frame:SetCallback("OnHide", function() AceGUI:ReleaseChildren(container)end)
	-- Fill Layout - the TabGroup widget will fill the whole frame
	frame:SetLayout("List")

	-- Create the TabGroup
	tab =  AceGUI:Create("TabGroup")
	tab:SetLayout("Flow")
	tab:SetRelativeWidth(1)
	tab:SetHeight(400)
	-- Setup which tabs to show
	tab:SetTabs({{text="Tab 1", value="tab1"}, {text="Tab 2", value="tab2"}})
	-- Register callback
	tab:SetCallback("OnGroupSelected", SelectGroup)
	-- Set initial Tab (this will fire the OnGroupSelected callback)
	tab:SelectTab("tab1")

	tab2 = AceGUI:Create("TabGroup")
	--tab2:SetPoint("BOTTOMLEFT", frame, "BOTTOMLEFT")
	tab2:SetRelativeWidth(1)
	tab2:SetLayout("Flow")
	tab2:SetTabs({{text="Tab 1", value="tab1"}, {text="Tab 2", value="tab2"}})
	-- Register callback
	tab2:SetCallback("OnGroupSelected", SelectGroup)
	-- Set initial Tab (this will fire the OnGroupSelected callback)
	tab2:SelectTab("tab1")
	-- add to the frame container
	frame:AddChild(tab)
	frame:AddChild(tab2)
	Slippy_Frame_Shown_12356741 = "1"
end

local slippy_options_frame = { -- Config Settings!!!!
	type = "group", 
	args = {
		SlippyOptionsDescription = {
			type = "description",
			name = "Easy way to view the materials and total estimated cost of user selected current enchants.\nFeel free to contact us at |cff00ff00SlippyAddon@gmail.com\r",
		},
		--Options Frame!
		SlippyOptionsHeader = {
			order = 2,
			type = "header",
			name = GetAddOnMetadata("Slippy", "title").." v"..GetAddOnMetadata("Slippy", "version"),
		},
		SlippyOptFrame = {
			order = 3,
			type  = "group",
			name  = "Enchant Options",
			handler = Slippy,
			args = {
				Header = {  -- Title which appears as a main-tab-option
					type = "header",
					order = 1,
					name = "Enchant Options"
				},
				description = {  -- Description - shows up at the top of the right-pane
					type = "description",
					order = 2,
					name = "Set values for materials in gold.\nExample: 20.32 = 20 Gold, 32 Silver",
				},
				infiniteDust = {
					type = "input",
					name = "Infinite Dust",
					desc = "The cost of 1 infinite dust in gold",
					get = "GetDust",
					set = "SetDust",
				},
				gce = {
					type = "input",
					name = "Greater Cosmic Essence",
					desc = "The cost of 1 greater cosmic essence in gold",
					get = "GetGCE",
					set = "SetGCE",
				},
				lce = {
					type = "input",
					name = "Lesser Cosmic Essence",
					desc = "The cost of 1 lesser cosmic essence in gold",
					get = "GetLCE",
					set = "SetLCE",
				},
				dream = {
					type = "input",
					name = "Dream Shard",
					desc = "The cost of 1 dream shard in gold",
					get = "GetDream",
					set = "SetDream",
				},
				abyss = {
					type = "input",
					name = "Abyss Crystal",
					desc = "The cost of 1 abyss crystal in gold",
					get = "GetAbyss",
					set = "SetAbyss",
				},
				titanium = {
					type = "input",
					name = "Titanium Bar",
					desc = "The cost of 1 Titanium Bar in gold",
					get = "GetTitanium",
					set = "SetTitanium",
				},
				water = {
					type = "input",
					name = "Crystallized Water",
					desc = "The cost of 1 Crystallized Water in gold",
					get = "GetWater",
					set = "SetWater",
				},
				earth = {
					type = "input",
					name = "Eternal Earth",
					desc = "The cost of 1 eternal earth in gold",
					get = "GetEarth",
					set = "SetEarth",
				},
				air = {
					type = "input",
					name = "Eternal Air",
					desc = "The cost of 1 eternal air in gold",
					get = "GetAir",
					set = "SetAir",
				},
				titansteel = {
					type = "input",
					name = "Titansteel Bar",
					desc = "The cost of 1 titansteel bar in gold",
					get = "GetTitansteel",
					set = "SetTitansteel",
				},
			},
		},
	},
}
	
function Slippy:Register_Options_Frame() 
	LibStub("AceConfig-3.0"):RegisterOptionsTable("Slippy", slippy_options_frame)
	self.optionsFrame = LibStub("AceConfigDialog-3.0"):AddToBlizOptions("Slippy", "Slippy")
end

-- Creates a frame with the proper offsets. Easy of coding purposes. 
-- Returns a reference to the created frame.
function Slippy:CreateFrame(anchor,anchorpoint,x,y,width,height,name,type1,buttonText)
	TempFrame=CreateFrame(type1,"SlippyGUI_MainBackground_Frame_"..name,anchor)
	if(TempFrame:IsVisible()) then
		while 1>0 do
			if(TempFrame:IsVisible()==nil) then
				break;
			end
			TempFrame:Hide()
		end
	end
	
	if(strlower(type1)=="button") then
		MyFrame=CreateFrame(type1,"SlippyGUI_MainBackground_Frame_"..name,anchor,"UIPanelButtonTemplate")
		MyFrame:SetText(buttonText)
	else
		MyFrame=CreateFrame(type1,"SlippyGUI_MainBackground_Frame_"..name,anchor)
	end
	
	MyFrame:SetPoint(anchorpoint,x,y)
	MyFrame:SetWidth(width)
	MyFrame:SetHeight(height)
	return MyFrame
end

-- the long and complicated function which handles ALL swapping of Main content for the main-frame. 
-- The basic gist of it is that it inputs a number which corresponds to a button on the left-menu. 
-- That number is then used to create a frame and anchor that frame to the main-content area.
function Slippy:SlippyGUI_SelectGroup(number)
	f=Slippy:CreateFrame(SlippyGUI_MainBackground1,"TOPLEFT",15,-15,555,310,"Test","Frame")
	g=Slippy:CreateFrame(f,"TOPLEFT",5,-5,100,20,"Button_1Test","Button","This is Some text")
end
	
	
	
	
	

data.lua
local twoH = {}		-- has 2 entries
twoH[0] = {"Greater Savagery", "Massacre"}		-- name of enchant
twoH[1] = {false, false}						-- status of checkbox - queued to be made? (true / false)
twoH[2] = {6, 40}								-- infinite dust
twoH[3] = {2, 6}								-- greater cosmic essences
twoH[4] = {0, 0}								-- lesser cosmic essences
twoH[5] = {0, 0}								-- dream shards
twoH[6] = {0, 6}								-- abyss crystals
twoH[7] = {0, 0}								-- titanium bar
twoH[8] = {0, 0}								-- crystallized water
twoH[9] = {0, 0}								-- eternal earth
twoH[10] = {0, 0}								-- eternal air
twoH[11] = {0, 0}								-- titansteel bar
twoH[12] = {0, 0}								-- cost
twoH[13] = {38992, 44463}						-- ItemID of corresponding scroll

local boots = {}		-- has 8 entries
boots[0] = {"Icewalker", "Greater Spirit", "Greater Fortitude", "Superior Agility", "Greater Assault", "Assault",
			"Tuskarr's Vitality", "Greater Vitality"}
boots[1] = {false, false, false, false, false, false, false, false}
boots[2] = {8, 10, 3, 16, 0, 4, 10, 8}
boots[3] = {0, 1, 0, 4, 4, 0, 2, 2}
boots[4] = {0, 0, 3, 0, 0, 4, 0, 0}
boots[5] = {0, 0, 0, 0, 4, 0, 2, 0}
boots[6] = {0, 0, 0, 0, 0, 0, 0, 0}
boots[7] = {0, 0, 0, 0, 0, 0, 0, 0}
boots[8] = {1, 0, 0, 0, 0, 0, 0, 0}
boots[9] = {0, 0, 0, 0, 0, 0, 0, 0}
boots[10] = {0, 0, 0, 0, 0, 0, 0, 0}
boots[11] = {0, 0, 0, 0, 0, 0, 0, 0}
boots[12] = {0, 0, 0, 0, 0, 0, 0, 0}
boots[13] = {38986, 38961, 38966, 38976, 44469, 44449, 39006, 38974}

local bracers = {}	-- has 9 entries
bracers[0] = {"Superior Spellpower", "Greater Spellpower", "Greater Assault", "Major Stamina", "Major Spirit", "Striking",
				"Exceptional Intellect", "Expertise", "Greater Stats"}
bracers[1] = {false, false, false, false, false, false, false, false, false}
bracers[2] = {6, 8, 24, 0, 16, 6, 10, 14, 16}
bracers[3] = {6, 0, 6, 4, 4, 0, 0, 3, 3}
bracers[4] = {0, 4, 0, 0, 0, 0, 0, 0, 0}
bracers[5] = {1, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[6] = {0, 0, 0, 1, 0, 0, 0, 0, 0}
bracers[7] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[8] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[10] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[11] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
bracers[13] = {44470, 38997, 44815, 44947, 38980, 38971, 38968, 38984, 38987}

local chest = {}		-- has 8 entries
chest[0] = {"Super Stats", "Powerful Stats", "Greater Defense", "Greater Mana Restoration", "Super Health", "Mighty Health",
			"Exceptional Mana", "Exceptional Resilience"}
chest[1] = {false, false, false, false, false, false, false, false}
chest[2] = {4, 0, 10, 4, 20, 0, 6, 0}
chest[3] = {0, 0, 0, 4, 4, 3, 0, 2}
chest[4] = {4, 0, 0, 0, 0, 0, 0, 0}
chest[5] = {0, 4, 0, 0, 0, 0, 0, 2}
chest[6] = {0, 4, 0, 0, 0, 0, 0, 0}
chest[7] = {0, 0, 0, 0, 0, 0, 0, 0}
chest[8] = {0, 0, 0, 0, 0, 0, 0, 0}
chest[9] = {0, 0, 0, 0, 0, 0, 0, 0}
chest[10] = {0, 0, 0, 0, 0, 0, 0, 0}
chest[11] = {0, 0, 0, 0, 0, 0, 0, 0}
chest[12] = {0, 0, 0, 0, 0, 0, 0, 0}
chest[13] = {38989, 44465, 39002, 38962, 39005, 38955, 38912, 38975}

local cloak = {}		-- has 9 entries
cloak[0] = {"Greater Speed", "Major Agility", "Mighty Armor", "Speed", "Titanweave", "Wisdom", "Superior Agility",
			"Spell Piercing", "Shadow Armor"}
cloak[1] = {false, false, false, false, false, false, false, false, false}
cloak[2] = {16, 8, 15, 6, 8, 0, 9, 15, 12}
cloak[3] = {4, 2, 2, 0, 0, 6, 0, 0, 0}
cloak[4] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
cloak[5] = {0, 2, 0, 0, 2, 0, 0, 0, 0}
cloak[6] = {0, 0, 0, 0, 0, 1, 0, 0, 1}
cloak[7] = {0, 0, 0, 0, 2, 0, 0, 0, 0}
cloak[8] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
cloak[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
cloak[10] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
cloak[11] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
cloak[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
cloak[13] = {39003, 44457, 39001, 44456, 38978, 39004, 38959, 38973, 38993}

local gloves = {}		-- has 7 entries
gloves[0] = {"Exceptional Spellpower", "Crusher", "Precision", "Armsman", "Major Agility", "Greater Assault", "Expertise"}
gloves[1] = {false, false, false, false, false, false, false}
gloves[2] = {4, 20, 0, 0, 0, 6, 12}
gloves[3] = {0, 4, 4, 0, 4, 1, 0}
gloves[4] = {1, 0, 0, 0, 0, 0, 0}
gloves[5] = {0, 1, 0, 2, 1, 0, 0}
gloves[6] = {0, 0, 0, 0, 0, 0, 0}
gloves[7] = {0, 0, 0, 0, 0, 0, 0}
gloves[8] = {0, 0, 0, 0, 0, 0, 0}
gloves[9] = {0, 0, 0, 8, 0, 0, 0}
gloves[10] = {0, 0, 0, 0, 0, 0, 0}
gloves[11] = {0, 0, 0, 0, 0, 0, 0}
gloves[12] = {0, 0, 0, 0, 0, 0, 0}
gloves[13] = {38979, 44458, 38953, 38990, 38967, 38964, 38951}

local shield = {} 	-- has 2 entries
shield[0] = {"Greater Intellect", "Defense"}
shield[1] = {false, false}
shield[2] = {12, 6}
shield[3] = {0, 0}
shield[4] = {0, 0}
shield[5] = {0, 0}
shield[6] = {0, 0}
shield[7] = {0, 0}
shield[8] = {0, 0}
shield[9] = {0, 6}
shield[10] = {0, 0}
shield[11] = {0, 0}
shield[12] = {0, 0}
shield[13] = {44455, 38954}

local staff = {}		-- has 2 entries
staff[0] = {"Greater Spellpower", "Spellpower"}
staff[1] = {false, false}
staff[2] = {40, 12}
staff[3] = {0, 2}
staff[4] = {0, 0}
staff[5] = {6, 0}
staff[6] = {6, 0}
staff[7] = {0, 0}
staff[8] = {0, 0}
staff[9] = {0, 0}
staff[10] = {0, 0}
staff[11] = {0, 0}
staff[12] = {0, 0}
staff[13] = {45056, 45060}

local weapon = {}		-- has 9 entries
weapon[0] = {"Blade Ward", "Exceptional Spellpower", "Mighty Spellpower", "Berserking", "Greater Potency",
			"Exceptional Agility", "Exceptional Spirit", "Accuracy", "Superior Potency"}
weapon[1] = {false, false, false, false, false, false, false, false, false}
weapon[2] = {0, 10, 30, 12, 2, 0, 16, 20, 10}
weapon[3] = {8, 2, 0, 4, 0, 0, 0, 4, 2}
weapon[4] = {0, 0, 0, 0, 4, 0, 0, 0, 0}
weapon[5] = {0, 0, 6, 4, 0, 4, 4, 4, 2}
weapon[6] = {4, 0, 6, 10, 0, 0, 0, 6, 4}
weapon[7] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
weapon[8] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
weapon[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
weapon[10] = {0, 0, 0, 0, 0, 4, 0, 0, 0}
weapon[11] = {1, 0, 0, 0, 0, 0, 0, 0, 0}
weapon[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0}
weapon[13] = {46026, 38991, 44467, 44493, 44453, 38995, 38963, 44497, 44466}

slippyData = {twoH, boots, bracers, chest, cloak, gloves, shield, staff, weapon}

core.lua
-- TODO LIST (in no particular order):
--	overall:
--		1. Make sure all of the included librarys are getting used and get rid of any that aren't.
--		2. Incorperate more with Quick Auctions: mat cost / scroll selling price.
--		3. Have the queue be refreshed when an enchant is cast and remove that enchant from the list.
--		4. Who doesn't love localizations?!
--		5. Make less stuff global!
--	gui.lua
--		1. Add tooltip functionality.
--		2. Add total costs of the materials to the summary window.

Slippy = LibStub("AceAddon-3.0"):NewAddon("Slippy", "AceConsole-3.0", "AceEvent-3.0")
AceGUI = LibStub("AceGUI-3.0")
Slippyversion = "1.5" -- current version of the addon
local QAAPI_STATUS

-- default values for the different material costs for new profiles in SlippyDB
local Slippydefaults = {
	profile = {
		infiniteDust = 2,
		gce = 15,
		lce = 5,
		dream = 6,
		abyss = 29,
		titanium = 18,
		water = 0.2,
		earth = 6,
		air = 21,
		titansteel = 130,
		vellum = true,
		totals = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
	},
}

-- Called when the addon is loaded
function Slippy:OnInitialize()
	--checks to make sure that the command QAAPI:GetData exists and warns the user if it doesn't.
	--will later automatically disable components that rely on the QAAPI to avoid errors.
	if pcall(function() QAAPI:GetData() end) then 
		QAAPI_STATUS = true
		self:Print("Loaded Slippy v" .. Slippyversion .. " and QAAPI successfully!")
	else
		self:Print("Loaded Slippy v" .. Slippyversion .. "!")
		self:Print("Warning: Your version of Quick Auctions does not support the QAAPI.")
		self:Print("Some features may be disabled or give errors.")
		self:Print("Please read the README file in the Slippy directory for more information.")
		QAAPI_STATUS = false
	end
	
	db = LibStub:GetLibrary("AceDB-3.0"):New("SlippyDB", Slippydefaults, true)
    self:RegisterChatCommand("sl", "ChatCommand")
    self:RegisterChatCommand("slippy", "ChatCommand")
	Slippy:Register_Options_Frame() -- Loads options-table
	Slippy_Frame_Shown_12356741 = "0"
	Slippy:FrameToggle()
	frame:Hide()
end

-- Called when the addon is enabled - for possible future use
function Slippy:OnEnable() end

-- Called when the addon is disabled - for possible future use
function Slippy:OnDisable() end

-- deals with slash commands
function Slippy:ChatCommand(input)
	if input == "config" then	-- '/sl config' opens up the options window
		InterfaceOptionsFrame_OpenToCategory(self.optionsFrame)
	elseif input == "summary" then	-- '/sl summary' opens up the main window
		frame:Show()
		tab:SelectTab("tab1")
	elseif input == "queue" then	-- '/sl queue' opens up the summary window
		Slippy:Calctotals()
		frame2:Show()
		tab2:SelectTab("tab21")
	elseif input == "test" then
		SlippyGUI_Background:Show()
	else	-- if the command is unrecognized, print out the slash commands to help the user
        self:Print("Slash Commands:")
		print("/sl summary - Opens window for viewing / selecting enchants.")
		print("/sl queue - Opens window for viewing totals / craft queue.")
		print("/sl config - Opens the options frame.")
    end
end

-- Calcuates the costs for all enchants of the passed slot.
function Slippy:Calc(slot)
	-- table.getn returns the size of the table which here is the number of enchants in a slot
	for i=1, table.getn(slippyData[slot][1]) do
		slippyData[slot][12][i] = slippyData[slot][2][i]*db.profile.dust
			+ slippyData[slot][3][i]*db.profile.gce + slippyData[slot][4][i]*db.profile.lce
			+ slippyData[slot][5][i]*db.profile.dream + slippyData[slot][6][i]*db.profile.abyss
			+ slippyData[slot][7][i]*db.profile.titanium + slippyData[slot][8][i]*db.profile.water
			+ slippyData[slot][9][i]*db.profile.earth + slippyData[slot][10][i]*db.profile.air
			+ slippyData[slot][11][i]*db.profile.titansteel
		--GUItable[2][slot][i]:SetText(slippydata[slot][12][i] .. "g")
	end
end

-- totals = dust, gce, lce, dream, abyss, titanium, water, earth, air, titansteel, wepeaonvellum, armorvellum, totalnumber
-- Calculates the total number of mats needed to craft the selected enchants
function Slippy:Calctotals()
	for i=1, 13 do	-- resets the totals to 0
		db.profile.totals[i] = 0
	end
	
	-- This series of for loops is where all the magic happens!
	-- The function BooleanToNumber converts the true/false value that represents
	-- if that enchant is selected or not to a 1 for true or a 0 for false.
	-- This is then multiplied by the number of the specific material needed for that enchant
	-- and added to the total of that specific material required for all selected enchants.
	-- Also, for each enchant selected, 1 is added to the number of the respective vellum needed
	-- to make an enchanting scroll for that enchant.
	for k=1, 5 do
		-- for j=1, 9 do
			-- for i=1, table.getn(slippyData[j][1]) do
				-- db.profile.totals[k] = db.profile.totals[k] + slippyData[j][k+1][i]*Slippy:BooleanToNumber(slippyData[j][1][i])
			-- end
		-- end
		for i=1, table.getn(slippyData[7][1]) do
			db.profile.totals[k] = db.profile.totals[k] + slippyData[8][k+1][i]*Slippy:BooleanToNumber(slippyData[8][1][i]) +
				slippyData[7][k+1][i]*Slippy:BooleanToNumber(slippyData[7][1][i]) +
				slippyData[1][k+1][i]*Slippy:BooleanToNumber(slippyData[1][1][i])
			if k == 1 then
				db.profile.totals[11] = db.profile.totals[11] + Slippy:BooleanToNumber(slippyData[8][1][i]) + Slippy:BooleanToNumber(slippyData[1][1][i])
				db.profile.totals[12] = db.profile.totals[12] + Slippy:BooleanToNumber(slippyData[7][1][i])
			end
		end
		
		for i=1, table.getn(slippyData[6][1]) do
			db.profile.totals[k] = db.profile.totals[k] + slippyData[6][k+1][i]*Slippy:BooleanToNumber(slippyData[6][1][i])
			if k == 1 then
				db.profile.totals[12] = db.profile.totals[12] + Slippy:BooleanToNumber(slippyData[6][1][i])
			end
		end
		
		for i=1, table.getn(slippyData[4][1]) do
			db.profile.totals[k] = db.profile.totals[k] + slippyData[2][k+1][i]*Slippy:BooleanToNumber(slippyData[2][1][i]) +
				slippyData[4][k+1][i]*Slippy:BooleanToNumber(slippyData[4][1][i])
			if k == 1 then
				db.profile.totals[12] = db.profile.totals[12] + Slippy:BooleanToNumber(slippyData[2][1][i]) +	Slippy:BooleanToNumber(slippyData[4][1][i])
			end
		end
		
		for i=1, table.getn(slippyData[5][1]) do
			db.profile.totals[k] = db.profile.totals[k] + slippyData[9][k+1][i]*Slippy:BooleanToNumber(slippyData[9][1][i]) +
				slippyData[5][k+1][i]*Slippy:BooleanToNumber(slippyData[5][1][i]) +
				slippyData[3][k+1][i]*Slippy:BooleanToNumber(slippyData[3][1][i])
			if k == 1 then
				db.profile.totals[11] = db.profile.totals[11] + Slippy:BooleanToNumber(slippyData[9][1][i])
				db.profile.totals[12] = db.profile.totals[12] + Slippy:BooleanToNumber(slippyData[5][1][i]) + Slippy:BooleanToNumber(slippyData[3][1][i])
			end
		end
	end
	-- these few commands deal with the extra materials (titanium bar, eternals, etc)
	db.profile.totals[6] = db.profile.totals[6] + 2*Slippy:BooleanToNumber(slippyData[5][1][5])
	db.profile.totals[7] = db.profile.totals[7] + Slippy:BooleanToNumber(slippyData[2][1][1])
	db.profile.totals[8] = db.profile.totals[8] + Slippy:BooleanToNumber(slippyData[4][1][3]) + 6*Slippy:BooleanToNumber(slippyData[7][1][2]) +
		8*Slippy:BooleanToNumber(slippyData[6][1][4])
	db.profile.totals[9] = db.profile.totals[9] + 4*Slippy:BooleanToNumber(slippyData[9][1][6])
	db.profile.totals[10] = db.profile.totals[10] + Slippy:BooleanToNumber(slippyData[9][1][1])
	
	-- calculates the total number of enchants selected by adding the weapon and armor vellum totals
	db.profile.totals[13] = db.profile.totals[12] + db.profile.totals[11]
end

-- simple function to convert true->1 and false->0
function Slippy:BooleanToNumber(value)
	if value then return 1
	else return 0
	end
end

-- resets all of the data when the "Reset Craft Queue" button is pressed
function Slippy:ResetData()
	queue[0]:SetText("You currently have no enchants in the queue.")
	self:Print("Craft Queue Reset")

	for slot=1,9 do
		for enchant=1, table.getn(slippyData[slot][1]) do
			slippyData[slot][1][enchant] = false
			slippyData[slot][12][enchant] = 0
		end
	end	
end

-- Allows an easy to use interface with the Quick Auctions API
-- The parameters are the slot for the enchant (boots, gloves, shield, etc)
-- and the index of the item (ex. slippyData[2][][#])
-- Returns a table with the following information (assuming you return into a variable called dataTbl):
-- dataTbl = {
    -- quantity = #, -- total number of items, including stacked
    -- onlyPlayer = true/false, -- Is the player the only one with items up
    -- records = { -- List of all items, by price per item. Multiple auctions at the same *per price* level and from the same person are merged
        -- {
            -- buyout = #, -- Buyout in copper, per item
            -- bid = #, -- Bid in copper, per item
            -- owner = "foo", -- Name 
            -- quantity = #, -- How many total are up at this tier/owner
            -- isPlayer = true/false,
        -- },
    -- },
-- }
function Slippy:GetQAData(slot, enchant)
	if not QAAPI_STATUS then 
		print("Error[Slippy:GetQAData]: No QA scan data found.")
	return nil end
	
	local sName, sLink = GetItemInfo(slippyData[slot][13][enchant])
	return QAAPI:GetData(sLink)
end

-- Parameters: Slot and Enchant number for slippyData[slot][#][enchant]
-- Returns the profit based off the lowest buyout price according to QA
function Slippy:CalcProfit(slot, enchant)
	local tbl = Slippy:GetQAData(slot, enchant)
	local profit
	
	if not pcall(function() return tbl.records[1].buyout end) then 
		self:Print("Error[Slippy:CalcProfit]: No QA scan data found.")
	return end
	
	profit = math.floor(tbl.records[1].buyout/10000 - slippyData[slot][12][enchant] + 0.5)
	return profit
end

-- converts an itemID into the name of the item. For example,
-- Slippy:GetItemName(slippyData[2][13][1]) returns "Scroll of Enchant Boots - Icewalker"
function Slippy:GetItemName(sItemID)
	local sName = GetItemInfo(sItemID)
	return sName
end